lib/puppet/provider/nameservice.rb

```diff
@@ -261,6 +269,10 @@ def groups
     groups.join(",")
   end
 
+  def groups
+    @groups ||= fetch_groups
```

Comment:
I was concerned that retrieving groups thrice would end up being slow, especially in the case when a direcotry service is involved... We can definitely pull out the caching commit and see if anybody complains, though :)